File: web/lib/crawl_rules.js
/** Bounded, deterministic rule configuration. No fetching or regex matchers. */
(function(root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory();
  else root.CrawlRules = factory();
}(typeof self !== 'undefined' ? self : this, function() {
  'use strict';
  const actions = ['discover', 'download', 'traverse'];
  const textFields = ['url_prefix', 'path_prefix', 'host', 'mime', 'kind'];
  const numberFields = ['min_depth', 'max_depth', 'min_bytes', 'max_bytes'];
  const object = value => value !== null && typeof value === 'object' && !Array.isArray(value);
  const keys = (value, allowed) => Object.keys(value).every(key => allowed.includes(key));
  function validateRules(config) {
    if (!object(config) || !keys(config,['version','rules']) || config.version !== 1 ||
        !Array.isArray(config.rules) || config.rules.length > 100) throw Error('Invalid rule configuration');
    const ids = new Set();
    for (const rule of config.rules) {
      if (!object(rule) || !keys(rule,['id','match','decision']) ||
          typeof rule.id !== 'string' || !/^[A-Za-z0-9_-]{1,64}$/.test(rule.id) || ids.has(rule.id)) throw Error('Invalid rule id');
      ids.add(rule.id);
      const match = rule.match, decision = rule.decision;
      if (!object(match) || !keys(match,[...textFields,...numberFields])) throw Error('Unknown matcher');
      if (!object(decision) || !Object.keys(decision).length || !keys(decision,actions) ||
          Object.values(decision).some(value=>typeof value !== 'boolean')) throw Error('Invalid decision');
      for (const [key,value] of Object.entries(match)) {
        if (numberFields.includes(key)) {
          if (!Number.isSafeInteger(value) || value < 0) throw Error('Invalid numeric bound');
        } else if (typeof value !== 'string' || !value || Array.from(value).length > 2048 || /[\x00-\x1f]/.test(value)) throw Error('Invalid matcher text');
      }
      for (const unit of ['depth','bytes']) {
        if ((match['min_'+unit] ?? 0) > (match['max_'+unit] ?? Number.MAX_SAFE_INTEGER)) throw Error('Reversed interval');
      }
      if ('kind' in match && !['page','asset','sitemap'].includes(match.kind)) throw Error('Invalid kind');
      if ('host' in match && !/^[a-z0-9]+(?:[.-][a-z0-9]+)*$/.test(match.host)) throw Error('Invalid host');
      if ('path_prefix' in match && !match.path_prefix.startsWith('/')) throw Error('Invalid path prefix');
      if ('url_prefix' in match) {
        const url = new URL(match.url_prefix);
        if (!['http:','https:'].includes(url.protocol) || !url.hostname || url.username || url.password) throw Error('Invalid URL prefix');
      }
    }
    return JSON.parse(JSON.stringify(config));
  }
  function evaluateRules(config, resource, inScope) {
    config = validateRules(config);
    if (typeof inScope !== 'boolean' || !object(resource) || typeof resource.url !== 'string' ||
        Array.from(resource.url).length > 8192 || /[\x00-\x1f]/.test(resource.url)) throw Error('Invalid resource');
    const url = new URL(resource.url);
    if (!['http:','https:'].includes(url.protocol) || !url.hostname || url.username || url.password ||
        !['page','asset','sitemap'].includes(resource.kind)) throw Error('Invalid resource URL or kind');
    for (const key of ['depth','bytes']) {
      if (resource[key] != null && (!Number.isSafeInteger(resource[key]) || resource[key]<0)) throw Error('Invalid resource count');
    }
    if (resource.mime != null && (typeof resource.mime !== 'string' || Array.from(resource.mime).length>2048)) throw Error('Invalid MIME');
    const result = {discover:inScope,download:inScope,traverse:inScope,rule_id:null,state:inScope?'decided':'out_of_scope',missing:[]};
    if (!inScope) return result;
    for (const rule of config.rules) {
      const missing = new Set();let mismatch = false;
      for (const [key,expected] of Object.entries(rule.match)) {
        if (numberFields.includes(key)) {
          const [direction,field]=key.split('_'),value=resource[field];
          if (value == null) missing.add(field);
          else if ((direction==='min' && value<expected) || (direction==='max' && value>expected)) mismatch=true;
        } else if (key==='mime') {
          if (resource.mime == null) missing.add('mime');
          else if (resource.mime.split(';')[0].trim().toLowerCase()!==expected.toLowerCase()) mismatch=true;
        } else {
          const value={url_prefix:resource.url,path_prefix:url.pathname || '/',host:url.hostname.toLowerCase(),kind:resource.kind}[key];
          if (!(key.endsWith('_prefix') ? value.startsWith(expected) : value===expected)) mismatch=true;
        }
      }
      if (mismatch) continue;
      result.rule_id=rule.id;
      if (missing.size) return {...result,download:false,traverse:false,state:'needs_metadata',missing:[...missing].sort()};
      Object.assign(result,rule.decision);
      result.download=result.discover && result.download;
      result.traverse=result.download && result.traverse;
      return result;
    }
    return result;
  }
  function scopeHosts(allowedHosts, aliases = {}) {
    const valid = value => typeof value==='string' && value.length<=253 && /^[a-z0-9]+(?:[.-][a-z0-9]+)*$/.test(value);
    if (!Array.isArray(allowedHosts) || allowedHosts.length>32 || !allowedHosts.every(valid) ||
        !object(aliases) || Object.keys(aliases).length>16) throw Error('Invalid host scope');
    const hosts=new Set(allowedHosts);
    for (const [primary,extra] of Object.entries(aliases)) {
      if (!allowedHosts.includes(primary) || !Array.isArray(extra) || extra.length>16 || !extra.every(valid)) throw Error('Invalid explicit alias');
      extra.forEach(host=>hosts.add(host));
    }
    if (hosts.size>32) throw Error('Too many hosts');
    return [...hosts].sort();
  }
  function sitemapLocations(text) {
    // Small strict XML subset also usable in the dependency-free test runtime.
    if (/[\x00-\x08\x0b\x0c\x0e-\x1f\ufffe\uffff]/.test(text)) throw Error('Invalid XML character');
    if (/<!DOCTYPE|<!ENTITY|<\/?[A-Za-z_][\w.-]*:/i.test(text)) throw Error('Unsupported sitemap XML');
    const structure={urlset:['url'],sitemapindex:['sitemap'],url:['loc','lastmod','changefreq','priority'],sitemap:['loc','lastmod'],loc:[],lastmod:[],changefreq:[],priority:[]};
    const stack=[],locations=[];let root=null,position=0,nodes=0,loc='';
    const decode=value=>value.replace(/&([^;]+);/g,(_,entity)=>{
      const entities={amp:'&',lt:'<',gt:'>',quot:'"',apos:"'"};
      if (Object.hasOwn(entities,entity)) return entities[entity];
      if (!/^#(?:x[0-9a-fA-F]+|[0-9]+)$/.test(entity)) throw Error('Invalid entity');
      const n=entity[1]==='x'?parseInt(entity.slice(2),16):Number(entity.slice(1));
      if (!Number.isSafeInteger(n) || (n<0x20 && ![9,10,13].includes(n)) || n===0xfffe || n===0xffff || n>0x10ffff || (n>=0xd800 && n<=0xdfff)) throw Error('Invalid entity');
      return String.fromCodePoint(n);
    });
    while (position<text.length) {
      if (text.startsWith('<!--',position)) {
        const end=text.indexOf('-->',position+4);if(end<0)throw Error('Malformed comment');position=end+3;continue;
      }
      if (text.startsWith('<?xml',position) && root===null) {
        const end=text.indexOf('?>',position+5);if(end<0)throw Error('Malformed declaration');position=end+2;continue;
      }
      if (text[position]!=='<') {
        let end=text.indexOf('<',position);if(end<0)end=text.length;
        const raw=text.slice(position,end);
        if (raw.replace(/&(?:amp|lt|gt|quot|apos|#x[0-9a-fA-F]+|#[0-9]+);/g,'').includes('&')) throw Error('Malformed entity');
        const value=decode(raw);
        if (stack.at(-1)==='loc') loc+=value;
        else if (!stack.length && value.trim()) throw Error('Text outside root');
        position=end;continue;
      }
      const end=text.indexOf('>',position);if(end<0)throw Error('Malformed tag');
      const token=text.slice(position,end+1);position=end+1;
      const closing=token.match(/^<\/([a-z]+)\s*>$/);
      if (closing) {
        if (stack.pop()!==closing[1]) throw Error('Mismatched tag');
        if (closing[1]==='loc' && loc.trim()) locations.push(loc.trim());
        if (locations.length>1000) throw Error('Sitemap location limit');
        continue;
      }
      const opening=token.match(/^<([a-z]+)((?:\s+[\w:.-]+\s*=\s*(?:"[^"<>]*"|'[^'<>]*'))*)\s*(\/?)>$/);
      if (!opening || !Object.hasOwn(structure,opening[1]) || ++nodes>6001) throw Error('Unsupported sitemap structure');
      const name=opening[1];
      if (opening[2].trim() && !/^\s+xmlns\s*=\s*(?:"[^"<>]*"|'[^'<>]*')\s*$/.test(opening[2])) throw Error('Unsupported attributes');
      const namespace=opening[2].match(/\sxmlns\s*=\s*["']([^"']*)["']/);
      if (namespace && namespace[1]!=='http://www.sitemaps.org/schemas/sitemap/0.9') throw Error('Unsupported namespace');
      if (!stack.length) {
        if (root || !['urlset','sitemapindex'].includes(name)) throw Error('Invalid root');root=name;
      } else if (!structure[stack.at(-1)].includes(name)) throw Error('Unsupported nesting');
      if (name==='loc') loc='';
      if (!opening[3]) stack.push(name);
    }
    if (stack.length || !root) throw Error('Unclosed sitemap');
    return {root,locations};
  }
  function discoverSitemap(text, sourceUrl, allowedHosts, visited=[]) {
    const hosts=scopeHosts(allowedHosts);
    if (typeof text!=='string' || new TextEncoder().encode(text).length>262144) throw Error('Sitemap byte limit');
    const identity=value=>{
      const url=new URL(value);
      if(!['http:','https:'].includes(url.protocol) || !url.hostname || url.username || url.password) throw Error('Invalid sitemap identity');
      url.hash='';return url.href;
    };
    const seen=new Set(visited.map(identity)),result={pages:[],sitemaps:[],excluded:[]};
    const source=new URL(sourceUrl);
    if (!['http:','https:'].includes(source.protocol) || !hosts.includes(source.hostname) || source.username || source.password) throw Error('Source outside scope');
    sourceUrl=identity(sourceUrl);
    if(seen.has(sourceUrl))return result;
    seen.add(sourceUrl);
    const {root,locations}=sitemapLocations(text);
    for(const location of locations) {
      if(Array.from(location).length>8192)throw Error('Sitemap URL limit');
      const url=new URL(location,sourceUrl);url.hash='';const value=url.href;
      if(!['http:','https:'].includes(url.protocol) || !hosts.includes(url.hostname) || url.username || url.password) {
        if(!result.excluded.includes(value))result.excluded.push(value);
      } else if(!seen.has(value)) {result[root==='sitemapindex'?'sitemaps':'pages'].push(value);seen.add(value);}
    }
    return result;
  }
  return {validateRules, evaluateRules, scopeHosts, discoverSitemap};
}));

File: cli/crawl_rules.py
"""Bounded, versioned rules. First matching rule wins; no regular expressions.

This module is deterministic and performs no network requests. Rules refine an
existing scope decision; they never grant access outside the caller's scope.
"""
import copy
import re
from urllib.parse import urlsplit, urljoin
import xml.etree.ElementTree as ET

MAX_RULES = 100
MAX_TEXT = 2048
MAX_INTEGER = 2 ** 53 - 1
ACTIONS = {'discover', 'download', 'traverse'}
TEXT_FIELDS = {'url_prefix', 'path_prefix', 'host', 'mime', 'kind'}
NUMBER_FIELDS = {'min_depth', 'max_depth', 'min_bytes', 'max_bytes'}


def validate_rules(config):
    """Validate a JSON rule configuration without modifying its input."""
    if (not isinstance(config, dict) or set(config) != {'version', 'rules'}
            or type(config['version']) is not int or config['version'] != 1
            or not isinstance(config['rules'], list) or len(config['rules']) > MAX_RULES):
        raise ValueError('Expected version 1 and at most 100 rules')
    identifiers = set()
    for rule in config['rules']:
        if not isinstance(rule, dict) or set(rule) != {'id', 'match', 'decision'}:
            raise ValueError('Rule requires id, match and decision')
        identifier = rule['id']
        if (not isinstance(identifier, str) or not re.fullmatch(r'[A-Za-z0-9_-]{1,64}', identifier)
                or identifier in identifiers):
            raise ValueError('Invalid or duplicate rule id')
        identifiers.add(identifier)
        match, decision = rule['match'], rule['decision']
        if not isinstance(match, dict) or set(match) - TEXT_FIELDS - NUMBER_FIELDS:
            raise ValueError('Unknown matcher; regular expressions are unsupported')
        if (not isinstance(decision, dict) or not decision or set(decision) - ACTIONS
                or any(type(value) is not bool for value in decision.values())):
            raise ValueError('Decisions must be boolean actions')
        for key, value in match.items():
            if key in NUMBER_FIELDS:
                if type(value) is not int or not 0 <= value <= MAX_INTEGER:
                    raise ValueError('Matcher bounds must be safe nonnegative integers')
            elif (not isinstance(value, str) or not value or len(value) > MAX_TEXT
                  or any(ord(char) < 32 for char in value)):
                raise ValueError('Invalid matcher text')
        for unit in ('depth', 'bytes'):
            if match.get('min_' + unit, 0) > match.get('max_' + unit, MAX_INTEGER):
                raise ValueError('Reversed matcher interval')
        if 'kind' in match and match['kind'] not in ('page', 'asset', 'sitemap'):
            raise ValueError('Unknown resource kind')
        if 'host' in match and not re.fullmatch(r'[a-z0-9]+(?:[.-][a-z0-9]+)*', match['host']):
            raise ValueError('Host must be an exact lowercase hostname')
        if 'path_prefix' in match and not match['path_prefix'].startswith('/'):
            raise ValueError('Path prefix must be absolute')
        if 'url_prefix' in match:
            parsed = urlsplit(match['url_prefix'])
            if parsed.scheme not in ('http', 'https') or not parsed.hostname or parsed.username or parsed.password:
                raise ValueError('URL prefix must be an HTTP URL without credentials')
    return copy.deepcopy(config)


def evaluate_rules(config, resource, *, in_scope):
    """Return first-match decisions; unknown response metadata holds downloads.

    A caller may acquire response headers for ``needs_metadata``, subject to its
    ordinary scope/politeness controls, but must reevaluate before reading bytes.
    """
    config = validate_rules(config)
    if type(in_scope) is not bool or not isinstance(resource, dict):
        raise ValueError('Explicit scope and resource required')
    url = resource.get('url')
    if not isinstance(url, str) or len(url) > 8192 or any(ord(c) < 32 for c in url):
        raise ValueError('Invalid resource URL')
    parsed = urlsplit(url)
    if parsed.scheme not in ('http', 'https') or not parsed.hostname or parsed.username or parsed.password:
        raise ValueError('Invalid resource URL')
    if resource.get('kind') not in ('page', 'asset', 'sitemap'):
        raise ValueError('Invalid resource kind')
    for key in ('depth', 'bytes'):
        value = resource.get(key)
        if value is not None and (type(value) is not int or not 0 <= value <= MAX_INTEGER):
            raise ValueError('Invalid resource count')
    if resource.get('mime') is not None and (not isinstance(resource['mime'], str) or len(resource['mime']) > MAX_TEXT):
        raise ValueError('Invalid resource MIME')
    result = dict(discover=in_scope, download=in_scope, traverse=in_scope,
                  rule_id=None, state='decided' if in_scope else 'out_of_scope', missing=[])
    if not in_scope:
        return result
    for rule in config['rules']:
        missing, mismatch = set(), False
        for key, expected in rule['match'].items():
            if key in NUMBER_FIELDS:
                direction, field = key.split('_', 1)
                value = resource.get(field)
                if value is None:
                    missing.add(field)
                elif (direction == 'min' and value < expected) or (direction == 'max' and value > expected):
                    mismatch = True
            elif key == 'mime':
                value = resource.get('mime')
                if value is None:
                    missing.add('mime')
                elif value.split(';')[0].strip().lower() != expected.lower():
                    mismatch = True
            else:
                value = {'url_prefix': url, 'path_prefix': parsed.path or '/',
                         'host': parsed.hostname.lower(), 'kind': resource['kind']}[key]
                if not (value.startswith(expected) if key.endswith('_prefix') else value == expected):
                    mismatch = True
        if mismatch:
            continue
        result['rule_id'] = rule['id']
        if missing:
            result.update(download=False, traverse=False, state='needs_metadata', missing=sorted(missing))
            return result
        result.update(rule['decision'])
        result['download'] = result['discover'] and result['download']
        result['traverse'] = result['download'] and result['traverse']
        return result
    return result


def scope_hosts(allowed_hosts, aliases=None):
    """Expand only operator-declared aliases, never redirect-derived hosts."""
    aliases = {} if aliases is None else aliases
    valid = lambda value: isinstance(value, str) and len(value) <= 253 and re.fullmatch(r'[a-z0-9]+(?:[.-][a-z0-9]+)*', value)
    if (not isinstance(allowed_hosts, list) or len(allowed_hosts) > 32
            or not all(valid(value) for value in allowed_hosts)
            or not isinstance(aliases, dict) or len(aliases) > 16):
        raise ValueError('Invalid host scope')
    hosts = set(allowed_hosts)
    for primary, extra in aliases.items():
        if primary not in allowed_hosts or not isinstance(extra, list) or len(extra) > 16 or not all(valid(value) for value in extra):
            raise ValueError('Invalid explicit alias')
        hosts.update(extra)
    if len(hosts) > 32:
        raise ValueError('Too many hosts')
    return sorted(hosts)


def discover_sitemap(text, source_url, allowed_hosts, visited=None):
    """Parse a bounded unprefixed sitemap subset, without fetching anything.

    Supports urlset/sitemapindex and the standard default namespace. Extensions,
    DTDs, CDATA, attributes other than the default namespace and nested location markup are rejected rather than partly inferred.
    The caller retains visited sitemap URLs across fetches to terminate cycles.
    """
    hosts = scope_hosts(allowed_hosts)
    if not isinstance(text, str) or len(text.encode('utf-8')) > 262144:
        raise ValueError('Sitemap byte limit')
    if re.search(r'<!DOCTYPE|<!ENTITY|<!\[CDATA\[|</?[A-Za-z_][\w.-]*:', text, re.I):
        raise ValueError('Unsupported sitemap XML declaration or prefix')
    seen = set(visited or [])
    result = {'pages': [], 'sitemaps': [], 'excluded': []}
    source = urlsplit(source_url)
    if source.scheme not in ('http', 'https') or source.hostname not in hosts or source.username or source.password:
        raise ValueError('Sitemap source outside scope')
    if source_url in seen:
        return result
    seen.add(source_url)
    try:
        # Input is a size-bounded Unicode string; all DTD/entity declarations
        # are rejected above before parsing, so entities cannot be defined.
        # Keep the pre-parser rejection regression test when changing this guard.
        root = ET.fromstring(text)  # nosec B314 - guarded, DTD-free sitemap subset
    except ET.ParseError as error:
        raise ValueError('Malformed sitemap XML') from error
    local = lambda tag: tag.removeprefix('{http://www.sitemaps.org/schemas/sitemap/0.9}')
    structure = {'urlset': {'url'}, 'sitemapindex': {'sitemap'},
                 'url': {'loc', 'lastmod', 'changefreq', 'priority'}, 'sitemap': {'loc', 'lastmod'},
                 'loc': set(), 'lastmod': set(), 'changefreq': set(), 'priority': set()}
    root_name = local(root.tag)
    if root_name not in ('urlset', 'sitemapindex'):
        raise ValueError('Unsupported sitemap root')
    nodes = list(root.iter())
    if len(nodes) > 6001:
        raise ValueError('Sitemap node limit')
    for node in nodes:
        name = local(node.tag)
        if node.attrib or name not in structure or any(local(child.tag) not in structure[name] for child in node):
            raise ValueError('Unsupported sitemap structure')
    locations = [node.text.strip() for node in nodes if local(node.tag) == 'loc' and node.text and node.text.strip()]
    if len(locations) > 1000:
        raise ValueError('Sitemap location limit')
    for location in locations:
        if len(location) > 8192:
            raise ValueError('Sitemap URL limit')
        url = urljoin(source_url, location)
        parsed = urlsplit(url)
        if parsed.scheme not in ('http', 'https') or parsed.hostname not in hosts or parsed.username or parsed.password:
            if url not in result['excluded']:
                result['excluded'].append(url)
        elif url not in seen:
            result['sitemaps' if root_name == 'sitemapindex' else 'pages'].append(url)
            seen.add(url)
    return result
